db/system_keyspace.cc

```diff
@@ -2620,6 +2620,13 @@ future<service::topology> system_keyspace::load_topology_state() {
             map = &ret.new_nodes;
         } else {
             map = &ret.transition_nodes;
+            // At most one node at a time can be in transitioning state.
```

Comment:
The point of it being a map is that if we will want, at some point, to support more then one operation at  a time we will not have to change all of the code to map since it will already assume that it is a map. By dropping the generic code and insert the assumptions that there is only one element in the map you removed all the benefit and left all the disadvantages.  I do not see the point in most of this patch. May be only the first check to make sure that there is only one node is there for now.

Reply:
OK, I will leave only the first check. It won't be needed after doing the changes from the comment below, but I think it's still a good idea because if we currently have more than one transitioning node, then something must have gone wrong.

I would also like to extend the comment above `transition_nodes` to something like "Currently at most one node at a time will be here, but the code shouldn't assume it because we might support more than one operation at a time in the future". It's not clear why this map is here and how it should be handled.

Reply:
> OK, I will leave only the first check. It won't be needed after doing the changes from the comment below, but I think it's still a good idea because if we currently have more than one transitioning node, then something must have gone wrong.

Yes, that what I wrote as well.

> 
> I would also like to extend the comment above `transition_nodes` to something like "Currently at most one node at a time will be here, but the code shouldn't assume it because we might support more than one operation at a time in the future". It's not clear why this map is here and how it should be handled.

OK.

Reply:
> The point of it being a map is that if we will want, at some point, to support more then one operation at a time

To be honest I don't think we will. Token ring changes? Will be quick and trivial once all data moves to tablets (we can skip streaming/repair altogether), so no need to parallelize there.

Reply:
Assuming a single transitioning node simplifies reasoning and code.

Reply:
> > The point of it being a map is that if we will want, at some point, to support more then one operation at a time
> 
> To be honest I don't think we will. Token ring changes? Will be quick and trivial once all data moves to tablets (we can skip streaming/repair altogether), so no need to parallelize there.

I also do not think (hope) that we will ever support parallel token ring changes, but we may be forced to support running in parallel several non token ring changing operations like rebuild.

Reply:
Since keeping the map might be useful, I did as I proposed earlier -- I left only the sanity check and extended the comment.